crates/subspace-networking/src/constructor.rs

```diff
@@ -506,16 +500,24 @@ where
                 ..ConnectedPeersConfig::default()
             }
         }),
-        autonat: (enable_autonat && external_addresses.is_empty()).then(|| AutonatConfig {
+        autonat: external_addresses.is_empty().then(|| AutonatConfig {
             use_connected: true,
             only_global_ips: !config.allow_non_global_addresses_in_dht,
             confidence_max: AUTONAT_MAX_CONFIDENCE,
             ..Default::default()
         }),
     });
 
-    if let KademliaMode::Static(mode) = kademlia_mode {
-        behaviour.kademlia.set_mode(Some(mode));
+    match (kademlia_mode, external_addresses.is_empty()) {
+        (KademliaMode::Static(mode), _) => {
+            behaviour.kademlia.set_mode(Some(mode));
+        }
+        (KademliaMode::Dynamic, false) => {
+            behaviour.kademlia.set_mode(Some(Mode::Server));
```

Comment:
We put too much trust in the provided address but ok for now.

Reply:
That is kind of the point, to trust what user says about their environment